## 목차 ##

- 오류 코드보다 예외를 사용하라
- Try-Catch-Finally 문부터 작성하라
- 미확인 예외를 사용하라
- 예외에 의미를 제공하라
- 호출자를 고려해 예외 클래스를 정의하라
- 정상 흐름을 정의하라
- null을 반환하지 마라
- null을 전달하지 마라

-------------------------------------------------------------

> 깨끗한 코드를 다루는 책에 오류처리를 논하는 장이 있어 이상하게 여길 수도 있지만<br/>
오류처리는 프로그램에 꼭 필요한 요소 중 하나이다.<br/>
무언가 잘못될 가능성은 늘 존재한다. 이를 바로잡을 책임 또한 우리 프로그래머에게 있다.  
<br/>

## 오류 코드보다 예외를 사용하라 ##

얼마 전까지만 해도 예외를 지원하지 않는 프로그래밍 언어가 많았다.  
예외 처리를 지원하지 않는 언어는 오류를 처리하고 보고하는데 제한적이다.  

```java
// 예외처리 코드가 빠진 코드
public class DeviceController {
  ...
  public void sendShutDown() {
    DeviceHandle handle = getHandle(DEV1);
    // Check the state of the device
    if (handle != DeviceHandle.INVALID) {
      // Save the device status to the record field
      retrieveDeviceRecord(handle);
      // If not suspended, shut down
      if (record.getStatus() != DEVICE_SUSPENDED) {
        pauseDevice(handle);
        clearDeviceWorkQueue(handle);
        closeDevice(handle);
      } else {
        logger.log("Device suspended. Unable to shut down");
      }
    } else {
      logger.log("Invalid handle for: " + DEV1.toString());
    }
  }
  ...
}
```
이러한 함수는 몇 가지 단점이 있다.

* 오류 확인을 잊어버리기 쉽다.   
* 함수를 호출한 즉시 오류를 확인해야 하 호출자 코드가 복잡해진다.   
<br/>
따라서 오류를 처리하기보단 아래 코드와 같이 예외를 던지는 편이 낫다.<br/>

```java
// 예외처리 코드가 포함 된 코드
public class DeviceController {
	  ...
	  public void sendShutDown() {
		    try {
			      tryToShutDown();
		    } catch (DeviceShutDownError e) {
			      logger.log(e);
		    }
	  }
	  private void tryToShutDown() throws DeviceShutDownError {
		    DeviceHandle handle = getHandle(DEV1);
		    DeviceRecord record = retrieveDeviceRecord(handle);
		    pauseDevice(handle); 
		    clearDeviceWorkQueue(handle); 
		    closeDevice(handle);
	  }
	  private DeviceHandle getHandle(DeviceID id) {
		    ...
		    throw new DeviceShutDownError("Invalid handle for: " + id.toString());
		    ...
	  }
	  ...
}
```
이러한 방법으로 코드를 작성하게 되면 여러 장점이 있다.<br />
* 오류코드와 논리가 분리되어 코드가 깔끔해진다.
* 가독성이 좋아진다.  
* 코드 품질이 나아진다.  
* 기능 알고리즘과 오류 처리 알고리즘이 분리할 수 있다.  
* 각 개념을 독립적으로 살펴보고 이해할 수 있게 된다.  

## Try-Catch-Finally 문부터 작성하라 ##

프로그램 안에다 범위를 지정한다는 사실은 매우 흥미롭다.

try 문은 트랜잭션과 비슷하다.<br/>
try 블록에서 무슨 일이 생기던 catch 블록에서 프로그램 상태를 일관성 있게 유지하기 때문이다.  
따라서 코드를 try-catch-finally 구조로 짜는 것이 좋다.<br/>
그러면 try 블록에서 무슨 일이 생기든지 호출자가 기대하는 상태를 정의하기 쉬워진다.<br/>
즉, 예외 처리를 지정함으로써 프로그래머는 코드의 방향성을 생각하면서 코드를 짤 수 있다.<br/>

테스트 코드를 통해 살펴보자

```java
// 테스트가 실패한다
@Test(expected = StorageException.class)
  public void retrieveSectionShouldThrowOnInvalidFileName() {
    sectionStore.retrieveSection("invalid - file");
  }
  
  public List<RecordedGrip> retrieveSection(String sectionName) {
    // dummy return until we have a real implementation
    return new ArrayList<RecordedGrip>();
  }
```

위 코드는 예외를 던지지 않아 단위 테스트가 실패한다

```java
// 이제 테스트는 통과한다.
  public List<RecordedGrip> retrieveSection(String sectionName) {
    try {
      FileInputStream stream = new FileInputStream(sectionName)
    } catch (Exception e) {
      throw new StorageException("retrieval error", e);
    }
  return new ArrayList<RecordedGrip>();
}
```

하지만 다음 코드는 예외를 던지므로 테스트가 성공해
리팩토링이 가능해진다.

```java
// 테스트도 통과하였고 Exception의 범위를 FileNotFoundException으로 줄여 정확히 어떤 Exception이 발생한지 알 수 있게된다.
  public List<RecordedGrip> retrieveSection(String sectionName) {
    try {
      FileInputStream stream = new FileInputStream(sectionName);
      stream.close();
    } catch (FileNotFoundException e) {
      throw new StorageException("retrieval error", e);
    }
    return new ArrayList<RecordedGrip>();
  }
```
3번째 코드에서는 2번째 검증 과정을 거쳐 어떤 Exception이 발생하였는지 확인이 가능해진다.

위 방법과 같이 예외를 일으키는 테스트 케이스를 작성한 후 테스트를 통과하게 코드를 작성하는 방법을 권장한다.</br>
이를 통해 범위 내에서의 트랜잭션 본질을 유지하기 쉬워진다.

## 미확인 예외를 사용하라 ##

확인된 예외를 사용할지 미확인 예외를 사용할지에 앞서 두 예외가 무엇인지 살펴보자

||Checked Exception|Unchecked Exception|   
|-|-----------------|-------------------|       
|처리 여부|반드시 예외를 처리해야 함|명시적인 처리를 강제하지 않음|  
|확인 시점|컴파일 단계|실행 단계|    
|예외 발생시 트랜잭션 처리|Roll-back 하지 않음|Roll-back 함|   
|대표 예외|Exception의 하위 클래스 중<br> Runtime Exception을 제외한 모든 예외<br> * IOException<br>* SQLException|Runtime Exception 하위 예외<br>* NullPointerException<br>* IllegalArgumentException<br>* IndexOutOfBoundException<br>* SystemException|  

프로그래머들은 확인된 예외의 장단점을 놓고 논쟁을 벌여왔다.  
하지만 안정적인 소프트웨어를 제작하는 요소로 꼭 필요한건 아니라는 사실이 분명해졌다.  
우리는 확인된 오류를 사용하기 위해 치르는 비용에 상응하는 이익을 제공하는지 따져봐야 한다.  

확인된(checked) 예외를 사용한 코드가 반드시 필요한 이유를 살펴보자

```java
public void catchFileInputException() {
    try { 
        a();
    } catch (FileInputException e) {
        throw new StorageException("retrieval error", e);
    }
}

public void a() throws FileInputException { 
    aa(); 
}

public void aa() throws FileInputException {
    aaa();
}

public void aaa() throws FileInputException {
    FileInputStream stream = new FileInputStream(sectionName);
    stream.close();
}
```

* 확인된(checked) 예외는 OCP를 위반한다.  
(OCP: 소프트웨어 개체는 확장에 대해 열려 있어야 하고, 수정에 대해서는 닫혀 있어야 한다'는 프로그래밍 원칙)
* 메소드에서 예외를 던졌는데 catch 문이 위에 있다면 모든 선언 부에 예외를 정의해야 한다.  
따라서 하위 단계에서 코드를 수정하면 모든 상위 단계 메소드 선언부를 고쳐야 한다.
* 모듈과 관련된 코드가 바뀌지 않아도 선언 부가 바뀌었으므로 모듈을 다시 빌드 후 배포 해야 한다.

때로는 확인된(checked) 예외도 유용하다.  
아주 중요한 라이브러리를 작성한다면 모든 예외를 잡아야 한다.  
하지만 일반적인 애플리케이션은 의존성이라는 비용이 이익보다 크기에 미확인 예외를 사용하는 것이 좋다.  

## 예외에 의미를 제공하라 ##

예외를 던질 때는 전후 상황을 충분히 덧붙여야 오류의 원인과 위치를 찾기 쉬워진다.  
자바가 예외에 호출 스택을 제공하지만 충분한 정보가 되지 않기 때문에  
오류 메시지에 정보를 담아 예외와 함께 던져야 한다. (ex 실패한 연산 이름, 실패 유형)  
애플리케이션이 로깅 기능을 사용한다면 catch 블록에서 오류를 기록하도록 충분한 정보를 남기는 것이 좋다.  

## 호출자를 고려해 예외 클래스를 정의하라 보완 ##

오류를 분류하는 방법은 수없이 많다. (위치, 유형 등등)
하지만 가장 중요한 관심사는 오류를 잡아내는 방법이 되어야 한다.

다음은 형편없이 분류한 사례이다.

```java
  // catch문의 내용이 비슷하다.
  
  ACMEPort port = new ACMEPort(12);
  try {
    port.open();
  } catch (DeviceResponseException e) {
    reportPortError(e);
    logger.log("Device response exception", e);
  } catch (ATM1212UnlockedException e) {
    reportPortError(e);
    logger.log("Unlock exception", e);
  } catch (GMXError e) {
    reportPortError(e);
    logger.log("Device response exception");
  } finally {
    ...
  }
```
위 코드는 외부 라이브러리가 던질 예외를 모두 잡아낸다.
하지만, 위 코드는 중복이 심하고 에러마다 기능을 추가해야 한다면 코드도 길어질 것이다.
즉, 시간이 지날수록 코드가 길고 복잡한 형태를 띠게 될 것이다.

따라서 아래 코드와 같이 호출하는 라이브러리 API를 감싸 예외 유형을 하나로 반환하면 좋다.

```java
LocalPort port = new LocalPort(12);
try{
  port.open();
  } catch (PortDeviceFailure e) {
    reportError(e);
    logger.log(e.getMessage(),e);
  } finally {
  ...
  }
  
  
public class LocalPort {
  private ACMEPort innerPort;
  
  public LocalPort(int portNumber){
    innerPort = new ACMEPort(portNumber);
    }
    
  public void open() {
    try{
      innerPort.open();
      } catch(DeviceResponseException e) {
        throw new PortDeviceFailure(e);
        } catch(ATM1212UnlockedException e) {
          throw new portDeviceFailure(e);
        } catch (GMXError e) {
          throw new ProtDeviceFailure(e);
        }
       }
       ...
     }
```

위 코드에서 LocalPort 클래스는 ACMEPort 클래스가 던지는 예외를 잡아 반환하는 감싸기 클래스일 뿐이다.

외부 API를 감싸서 사용함으로써 외부 라이브러리와 프로그램 사이에서 의존성을 줄이고 예외 유형 하나를 반환할 수 있게 된다.

## 정상 흐름을 정의하라 ##

위에서 설명한 지침을 따른다면 깔끔한 코드가 나온다.  
하지만 예외 처리 하는 것이 적절하지 않을 때도 있다.

예외 처리 또한 가독성을 낮추는 작업이기에 예외를 발생시키지 않는 코드를 작성하는 것이 좋다.  
즉, 예외 처리에 집중이 되어 보다 좋은 코드를 작성할 기회를 놓치지 않길 바란다.  
예외를 발생시키는 것보다 특수사례 패턴(클래스를 만들거나 객체를 조작해 특수 사례를 처리하는 방식)을 
이용해 예외적인 상황을 미연에 방지할 수 있다.  

## null을 반환하지 마라 ##

오류처리에 관해 이야기 할 때 우리가 오류를 유발하는 행위에 대해서도 논해야 한다.  
그중 첫 번째가 null을 반환하는 습관이다.  
한 줄 건너 하나씩 null을 확인하는 애플리케이션이 있다고 치자  
이런 코드는 나쁜 코드이다.  
이러한 코드는 일거리를 늘릴 뿐만 아니라 호출자에게 문제를 떠넘긴다  
애플리케이션 어디선가 날린 NullpointerException을 처리하기란 쉽지 않다.  

이러한 코드는  null 확인이 누락된 문제라 생각할 수도 있지만  
실제로는 null 확인이 너무 많아 발생한 문제이다.  


## null을 전달하지 마라 ##

null을 반환하는 메소드보다 메소드로 null 을 전달하는 방식이 더 나쁘다.  
정상적인 인수로 null을 기대하는 API가 아니라면 메소드로 null을 전달하는 코드는 최대한 피해야 한다.  
일반적으로 대다수의 프로그래밍 언어들은 파라미터로 들어온 null에 대해 적절한 방법을 제공하지 못한다.  
가장 이상적인 해법은 null을 파라미터로 받지 못하게 하는 것이다.  

## 결론 ##
깨끗한 코드는 읽기도 좋아야 하지만 안정성도 높아야 한다.  
이 둘은 상충하는 목표가 아니다.  
오류처리 코드를 프로그램 논리와 분리해 독자적인 사안으로 고려해야 하며  
이를 통해 독립적인 추론이 가능하게 하고 유지보수성도 높아진다.

